File: SIC_Backend/Repositories/SectionsRepository.cs
using Microsoft.EntityFrameworkCore;
using SIC_Backend.Data;
using SIC_Backend.Data.DTOs;
using SIC_Backend.Data.Models;

namespace SIC_Backend.Repositories
{
    public class SectionsRepository(ApplicationDbContext dbContext) : ISectionsRepository
    {
        public SectionDTO GetSectionById(int id)
        {
            var section = dbContext.Sections.Where(s => s.Id == id)
                .Include(s => s.Devices)
                .Select(SectionDTO.FromSection)
                .First();

            return section;
        }

        public IEnumerable<SectionDTO> GetSectionsByUserId(string userId)
        {
            return dbContext.Sections.Where(s => s.UserId == userId)
                .Include(s => s.Devices)
                .Select(SectionDTO.FromSection)
                .ToList();
        }

        public IEnumerable<SectionDTO> GetSectionsByPlaceId(int placeId)
        {
            return dbContext.Sections.Where(s => s.PlaceId == placeId)
                .Include(s => s.Devices)
                .Select(SectionDTO.FromSection)
                .ToList();
        }

        public async Task CreateSectionAsync(CreateSectionModel model)
        {
            var section = Section.FromCreateModel(model);

            dbContext.Sections.Add(section);

            await dbContext.SaveChangesAsync();
        }

        public async Task UpdateSectionAsync(int id, UpdateSectionModel model)
        {
            var section = await dbContext.Sections.FindAsync(id);

            if (section == null)
            {
                //throw new NotFoundException("Section not found");
            }

            section!.Name = model.Name;
            section.PlaceId = model.PlaceId;

            await dbContext.SaveChangesAsync();
        }

        public async Task DeleteSectionAsync(int id)
        {
            var section = await dbContext.Sections.FindAsync(id);

            if (section == null)
            {
                //throw new NotFoundException("Section not found");
            }

            dbContext.Sections.Remove(section!);
            await dbContext.SaveChangesAsync();
        }
    }
}

File: sic_home/lib/repositories/sections_repository.dart
import 'package:dio/dio.dart';
import 'package:sic_home/apis/sections_api.dart';
import 'package:sic_home/models/create_section_model.dart';
import 'package:sic_home/models/section.dart';
import 'package:sic_home/models/update_section_model.dart';
import 'package:sic_home/repositories/interfaces/i_sections_repository.dart';

class SectionsRepository implements ISectionsRepository {
  final SectionsApi _sectionsApi = SectionsApi(Dio());

  @override
  Future<void> createSection(CreateSectionModel section) async {
    return await _sectionsApi.createSection(section);
  }

  @override
  Future<void> deleteSection(int id) async {
    return await _sectionsApi.deleteSection(id);
  }

  @override
  Future<Section> getSectionById(int id) async {
    return await _sectionsApi.getSection(id);
  }

  @override
  Future<List<Section>> getSectionsByPlace(int placeId) async {
    return await _sectionsApi.getSectionsByPlace(placeId);
  }

  @override
  Future<List<Section>> getSectionsByUser(String userId) async {
    return await _sectionsApi.getSectionsByUser(userId);
  }

  @override
  Future<void> updateSection(int id, UpdateSectionModel section) async {
    return await _sectionsApi.updateSection(id, section);
  }
}
